larreco/QuadVtx/EvalVtx_module.cc

```diff
@@ -0,0 +1,135 @@
+
+protected:
+  std::string fTruthLabel;
+  std::vector<std::string> fVertexLabels;
+
+  TTree* fTree;
+};
+
+DEFINE_ART_MODULE(EvalVtx)
+
+// ---------------------------------------------------------------------------
+EvalVtx::EvalVtx(const fhicl::ParameterSet& pset) :
+  EDAnalyzer(pset),
+  fTruthLabel(pset.get<std::string>("TruthLabel")),
+  fVertexLabels(pset.get<std::vector<std::string>>("VertexLabels"))
+{
+  art::ServiceHandle<art::TFileService> tfs;
+
+  fTree = tfs->make<TTree>("vtxs", "vtxs");
+
+  fTree->Branch("evt", &gEvt);
+  fTree->Branch("true_x", &gTrueX);
+  fTree->Branch("true_y", &gTrueY);
+  fTree->Branch("true_z", &gTrueZ);
+
+  gRecoX.resize(fVertexLabels.size());
+  gRecoY.resize(fVertexLabels.size());
+  gRecoZ.resize(fVertexLabels.size());
+
+  for(unsigned int i = 0; i < fVertexLabels.size(); ++i){
+    const std::string& l = fVertexLabels[i];
+    fTree->Branch((l+"_x").c_str(), &gRecoX[i]);
+    fTree->Branch((l+"_y").c_str(), &gRecoY[i]);
+    fTree->Branch((l+"_z").c_str(), &gRecoZ[i]);
+  }
+}
+
+// ---------------------------------------------------------------------------
+recob::Vertex GetFirstVertex(const std::string& label, const art::Event& evt)
+{
+  art::Handle<std::vector<recob::Vertex>> vtxs;
+  evt.getByLabel(label, vtxs);
+
+  if(vtxs->empty()) return recob::Vertex();
+
+  return (*vtxs)[0];
+}
+
+// ---------------------------------------------------------------------------
+recob::Vertex GetVtxByAssns(const std::string& label, const art::Event& evt)
+{
+  art::Handle<std::vector<recob::Vertex>> vtxs;
+  evt.getByLabel(label, vtxs);
+
+  if(vtxs->empty()) return recob::Vertex();
+
+  art::Handle<std::vector<recob::PFParticle>> parts;
+  evt.getByLabel(label, parts);
+
+  art::FindManyP<recob::Vertex> fm(parts, evt, label);
+
+  for(unsigned int i = 0; i < parts->size(); ++i){
+    const int pdg = abs((*parts)[i].PdgCode());
+    if(pdg == 12 || pdg == 14 || pdg == 16){
+      const std::vector<art::Ptr<recob::Vertex>> vtxs = fm.at(i);
+      if(vtxs.size() == 1) return *vtxs[0];
+
+      if(vtxs.empty()) std::cout << "Warning: vertex list empty!" << std::endl;
+      if(!vtxs.empty()) std::cout << "Warning: " << vtxs.size() << " vertices for daughter?" << std::endl;
+    }
+  }
+
+  return recob::Vertex();
+}
+
+// ---------------------------------------------------------------------------
+void EvalVtx::analyze(const art::Event& evt)
+{
+  art::Handle<std::vector<simb::MCTruth>> truths;
+  evt.getByLabel(fTruthLabel, truths);
+  if(truths->empty()) return;
+
+  const simb::MCParticle& nu = (*truths)[0].GetNeutrino().Nu();
+
+  gEvt = evt.event();
+
+  gTrueX = nu.Vx();
+  gTrueY = nu.Vy();
+  gTrueZ = nu.Vz();
+
+  for(unsigned int i = 0; i < fVertexLabels.size(); ++i){
+    const recob::Vertex vtx = (fVertexLabels[i] == "pandora") ? GetVtxByAssns(fVertexLabels[i], evt) : GetFirstVertex(fVertexLabels[i], evt);
+    const recob::Vertex::Point_t reco_vtx = vtx.position();
```

Comment:
The trouble is that the output tree is structured to allow comparison of vertexers event by event. If one vertexer found a vertex and another didn't, this is important information. You can't just drop the whole event. My plotting scripts know what a default-valued vertex means, but it would be possible to add an isValid branch to the tree to make it more explicit, if that's desirable.

Reply:
I see.  If downstream code already depends on this setup, then it's not worth redoing.  You might consider adding a comment as to why this is being done this way, but that's up to you.  Feel free to resolve this comment according to your liking.

Reply:
Added a comment.